src/qt/addresstablemodel.cpp

```diff
@@ -369,17 +370,14 @@ QString AddressTableModel::addRow(const QString &type, const QString &label, con
     {
         // Generate a new address to associate with given label
         CPubKey newKey;
-        if(!wallet->GetKeyFromPool(newKey))
-        {
+        if (!wallet->GetKeyFromPool(newKey) || (wallet->IsHDEnabled() && !wallet->HasUnusedKeys(GetArg("-keypoolcritical", DEFAULT_KEYPOOL_CRITICAL)))) {
```

Comment:
In commit "[wallet] add checks for keypool critical"

It seems like a less repetitive (and safer?) approach would be to make GetKeyFromPool return false when number of keys is below critical, and maybe have an override if needed.